### DISCUSSION TITLE
useQuery causes unnecessary re-renders

### DISCUSSION THREAD

**Discussion by mapsgeek** [urql-graphql/urql#1707] [Q&A] [answered]
Hi, i'm new to React and not sure if this is a normal behavior but here's an example of the issue i want to ask about
https://codesandbox.io/s/urql-rerendering-dwrme

it's just a simple React-router setup with 2 pages and one page fetches some data but when you navigate back and forth
why-did-you-render will throw warnings about unnecessary renders, yet i'm not sure if this should be a bug or an expected behavior or how much should i worry about unnecessary renders and optimizing them?

**Comment by kitten** [comment, marked as answer]
First of all, the notion and idea of "unnecessary re-renders" is mostly not up-to-date and in line with modern React. In fact, with concurrent mode being on the horizon a lot of changes indirectly require us to re-render more often. Most commonly, what you'll see is a single re-render, namely:

1. Your component mounts and `useQuery` computes its initial state
2. The effect runs and receives the "up-to-date" state (or `fetching: true`)
3. If the hook is fetching it'll in the future receive a first result and that's when `fetching: false` will also be set.

in other words, you're seeing this because we actively avoid rendering stale state on mount. If you were to look what's going on internally, in `urql`'s `Client`, you'd also see that two queries are started. The first query is only used to compute the mounting state and is aborted immediately. This is equivalent to `client.readQuery`. It ensures that the cache has a chance to provide you with an initially correct state.

The effect-based render then restarts the query and this will ensure that the actual client-logic runs, which means this is what will send the API request, receive future cache updates, etc. But it's also guaranteed to give you a fresh result.

That's because overall it's cheaper for us to trigger state updates, rather than to eagerly always deep-diff and deep-compare results. Deeply comparing the result is a dangerous practice as it may outpace the cost of letting React (and hence the virtual DOM) update on its own, which is in itself also like a deep-diffing process.

You can interrupt this cycle by strategically placing `React.memo` around your deeper, expensive components. Assuming that the query is hoisted as far as possible in your tree, to a high-up page-like component, then your memos can help to avoid expensive renders. That's assuming you'd only start dropping these in though once you start to see _actual_ performance issues.

We've talked to members of the React team to ensure that `useQuery` (and our other hook implementations) are not only React Concurrent-compatible, but also avoid tearing and other common issues in React. So you can rest assured that as long as your app works, and you don't see expensive renders, there's nothing you have to do. In fact, in the future version of React if you choose to use Concurrent rendering, you may even see these renders stop being an issue, if you do run into any problems. But that said, this is already unlikely early in development.

**Reply by mapsgeek** [reply]
Thank you for the explanation and as i said I'm new to react and yes honestly the concept of tracking of renders is quite tedious and time consuming process for a new comer as i come from vue as far as i learned and coded with it i didn't have to worry about any of this or even knew about it versus in react i don't know if it's due a bad structure in my code or a normal behavior.
your answer also explains why this doesn't happen when you specify the request-policy to `network-only` .
